cdap-runtime-ext-dataproc/src/main/java/io/cdap/cdap/runtime/spi/runtimejob/DataprocRuntimeJobManager.java

```diff
@@ -330,8 +335,16 @@ private LocalFile uploadFile(String bucket, String targetFilePath,
       LOG.debug("File's Location type : {} and Location : {}. ", bucketObj.getLocationType(), bucketObj.getLocation());
     }
     try (InputStream inputStream = openStream(localFile.getURI());
-         WriteChannel writer = getStorageClient().writer(blobInfo)) {
+         WriteChannel writer = getStorageClient().writer(blobInfo,
+                                                         Storage.BlobWriteOption.doesNotExist())) {
```

Comment:
Is it `ifGenerationMatch`? We should not just skip upload if file exists. Are we 100% sure it's impossible to have partial / 0-size uploads? It's much safer to overwrite in this case

Reply:
Yeah, it is `ifGenerationMatch=0` for new file uploads. I think object is only finalised when the upload is completed. So, if it gets successfully uploaded in the previous attempt than the precondition will fail in the second attempt, hence I ignored `412`. 

Also `storage.writer` will create a resumable session, the object won't actually "exist" until the resumable upload is finalized AFAIK, so I don't think there should be worry about partial uploads.

Reply:
Well, the "upload completed" is indicated any time you call "writer.close". This would happen automatically any time you exit the `try` block, e.g. if you had any problems reading (and there is an exception) you would write a partial file.

Reply:
Also writer javadoc says "data may not be visible" until close is called, so in some case data may be visible even before